### PR TITLE
treewide: Clean up multi-arch tools and usage

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

See the commit messages for more details, but the principles basically are:

  - Prefer custom stdenvs to tools as extra deps
  - Don't needlessly force gcc or clang
  - Dispatch on more appropriate conditions

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-33087

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -5603,11 +5597,11 @@ with pkgs;
...
   gcc-unwrapped = gcc.cc;
 
-  gccStdenv = if (!stdenv.isDarwin) then stdenv else stdenv.override {
+  gccStdenv = if stdenv.cc.isGNU then stdenv else stdenv.override {
     allowedRequisites = null;
     cc = gcc;
-    # Include unwrapped binaries like AS, etc. and remove libcxx/libcxxabi
-    extraBuildInputs = [ stdenv.cc.cc ];
+    # Remove libcxx/libcxxabi
+    extraBuildInputs = [ ];
```
Comment (LnL7) [inline pkgs/top-level/all-packages.nix]: This breaks `tests.cc-wrapper-gcc` on darwin, but there's probably a nicer way to fix it.

**Review by LnL7 — APPROVED** [review]
looks good

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/top-level/all-packages.nix]: I did the same method, but made it darwin-specific and specifically clang.

**Pull request merged**